File: CS/MyTokenEdit.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Popup;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;

namespace WindowsFormsApplication2 {
    [UserRepositoryItem("RegisterMyTokenEdit")]
    public class RepositoryItemMyTokenEdit : RepositoryItemTokenEdit {
        static RepositoryItemMyTokenEdit() {
            RegisterMyTokenEdit();
        }
        public const string CustomEditName = "MyTokenEdit";

        public RepositoryItemMyTokenEdit() {
        }
        public override string EditorTypeName {
            get { return CustomEditName; }
        }
        public static void RegisterMyTokenEdit() {
            Image img = null;
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(CustomEditName, typeof(MyTokenEdit), typeof(RepositoryItemMyTokenEdit), typeof(TokenEditViewInfo), new TokenEditPainter(), true, img));
        }
    }

    [ToolboxItem(true)]
    public class MyTokenEdit : TokenEdit {
        static MyTokenEdit() {
            RepositoryItemMyTokenEdit.RegisterMyTokenEdit();
        }
        protected override TokenEditPopupForm CreatePopupForm() {
            return new MyTokenEditPopupForm(this);
        }
        protected override bool CanAddNewToken() {
            return true;
        }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMyTokenEdit Properties {
            get { return base.Properties as RepositoryItemMyTokenEdit; }
        }
        public override string EditorTypeName {
            get { return RepositoryItemMyTokenEdit.CustomEditName; }
        }
    }

    public class MyTokenEditPopupForm : TokenEditPopupForm {
        public MyTokenEditPopupForm(TokenEdit tokenEdit)
            : base(tokenEdit) {
        }
        protected override IList GetDataSource() {
            TokenEditTokenCollection tokCol = Properties.Tokens;
            TokenEditSelectedItemCollection selCol = Properties.SelectedItems;
            List<TokenEditToken> list = new List<TokenEditToken>(Properties.Tokens.Count);
            for(int i = 0; i < tokCol.Count; i++) {
                TokenEditToken tok = tokCol[i];
                if(IsSelected(tok)) {
                    ObjectWrapper wr = (ObjectWrapper)tok.Value;
                    tok = new TokenEditToken(tok.Description, wr.Clone());
                }
                list.Add(tok);
            }
            return list;
        }
        protected bool IsSelected(TokenEditToken tok) {
            foreach(TokenEditToken item in Properties.SelectedItems) {
                if(object.ReferenceEquals(tok, item)) return true;
            }
            return false;
        }
    }

    public class ObjectWrapper : ICloneable {
        int obj;
        public ObjectWrapper(int obj) {
            this.obj = obj;
        }
        public override string ToString() {
            return Obj.ToString();
        }
        public int Obj { get { return obj; } }

        object ICloneable.Clone() {
            return Clone();
        }
        public ObjectWrapper Clone() {
            return new ObjectWrapper(Obj);
        }
    }
}

File: VB/MyTokenEdit.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Popup
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo

Namespace WindowsFormsApplication2
    <UserRepositoryItem("RegisterMyTokenEdit")> _
    Public Class RepositoryItemMyTokenEdit
        Inherits RepositoryItemTokenEdit

        Shared Sub New()
            RegisterMyTokenEdit()
        End Sub
        Public Const CustomEditName As String = "MyTokenEdit"

        Public Sub New()
        End Sub
        Public Overrides ReadOnly Property EditorTypeName() As String
            Get
                Return CustomEditName
            End Get
        End Property
        Public Shared Sub RegisterMyTokenEdit()
            Dim img As Image = Nothing
            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(CustomEditName, GetType(MyTokenEdit), GetType(RepositoryItemMyTokenEdit), GetType(TokenEditViewInfo), New TokenEditPainter(), True, img))
        End Sub
    End Class

    <ToolboxItem(True)> _
    Public Class MyTokenEdit
        Inherits TokenEdit

        Shared Sub New()
            RepositoryItemMyTokenEdit.RegisterMyTokenEdit()
        End Sub
        Protected Overrides Function CreatePopupForm() As TokenEditPopupForm
            Return New MyTokenEditPopupForm(Me)
        End Function
        Protected Overrides Function CanAddNewToken() As Boolean
            Return True
        End Function
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
        Public Shadows ReadOnly Property Properties() As RepositoryItemMyTokenEdit
            Get
                Return TryCast(MyBase.Properties, RepositoryItemMyTokenEdit)
            End Get
        End Property
        Public Overrides ReadOnly Property EditorTypeName() As String
            Get
                Return RepositoryItemMyTokenEdit.CustomEditName
            End Get
        End Property
    End Class

    Public Class MyTokenEditPopupForm
        Inherits TokenEditPopupForm

        Public Sub New(ByVal tokenEdit As TokenEdit)
            MyBase.New(tokenEdit)
        End Sub
        Protected Overrides Function GetDataSource() As IList
            Dim tokCol As TokenEditTokenCollection = Properties.Tokens
            Dim selCol As TokenEditSelectedItemCollection = Properties.SelectedItems
            Dim list As New List(Of TokenEditToken)(Properties.Tokens.Count)
            For i As Integer = 0 To tokCol.Count - 1
                Dim tok As TokenEditToken = tokCol(i)
                If IsSelected(tok) Then
                    Dim wr As ObjectWrapper = CType(tok.Value, ObjectWrapper)
                    tok = New TokenEditToken(tok.Description, wr.Clone())
                End If
                list.Add(tok)
            Next i
            Return list
        End Function
        Protected Function IsSelected(ByVal tok As TokenEditToken) As Boolean
            For Each item As TokenEditToken In Properties.SelectedItems
                If Object.ReferenceEquals(tok, item) Then
                    Return True
                End If
            Next item
            Return False
        End Function
    End Class

    Public Class ObjectWrapper
        Implements ICloneable


        Private obj_Renamed As Integer
        Public Sub New(ByVal obj As Integer)
            Me.obj_Renamed = obj
        End Sub
        Public Overrides Function ToString() As String
            Return Obj.ToString()
        End Function
        Public ReadOnly Property Obj() As Integer
            Get
                Return obj_Renamed
            End Get
        End Property

        Private Function ICloneable_Clone() As Object Implements ICloneable.Clone
            Return Clone()
        End Function
        Public Function Clone() As ObjectWrapper
            Return New ObjectWrapper(Obj)
        End Function
    End Class
End Namespace
